File: problem_sets/w2/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int computeGrade(string s);
void calculateGrade(int index);

int main(void)
{
    // prompt the user for the text
    string text = get_string("Text: ");

    // compute the text grade based on the Coleman-Liau index
    int index = computeGrade(text);

    // print the grade
    calculateGrade(index);
}

int computeGrade(string s)
{
    int words = 0;
    int letters = 0;
    int sentences = 0;

    for (int i = 0, n = strlen(s); i <= n; i++)
    {
        if (isspace(s[i]) || s[i] == '\0')
        {
            words++;
        }
        else if (isalpha(s[i]))
        {
            letters++;
        }
        else if ((s[i] == '.' || s[i] == '!' || s[i] == '?') &&
                 (i > 0 && s[i - 1] != '!' && s[i - 1] != '?'))
        {
            sentences++;
        }
    }

    float L = letters / (words / 100.0);
    float S = sentences / (words / 100.0);

    int index = round(0.0588 * L - 0.296 * S - 15.8);

    return index;
}

void calculateGrade(int index)
{
    if (index < 0)
    {
        printf("Before Grade 1\n");
    }
    else if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }
}

File: problem_sets/w6/sentimental-readability/readability.py
from cs50 import get_string


text = get_string("Text: ")

letters = 0
words = 0
sentences = 0


# Count the number of letters and sentences
for c in text:
    if c.isalpha():
        letters += 1
    elif c in [".", "!", "?"]:
        sentences += 1

# Count the number of words
for word in text.split():
    words += 1


# Calculate the Coleman-Liau index
# L is the average number of letters per 100 words in the text
L = 100 * letters / words

# S is the average number of sentences per 100 words in the text
S = sentences / (words / 100)

CLindex = round(0.0588 * L - 0.296 * S - 15.8)

# Print the grade level
if CLindex < 0:
    print("Before Grade 1")
elif CLindex >= 16:
    print("Grade 16+")
else:
    print(f"Grade {CLindex}")
